src/rust/engine/graph/src/lib.rs

```diff
@@ -543,6 +574,28 @@ impl<N: Node> Graph<N> {
     }
   }
 
+  fn detect_cycle(
+    src_id: EntryId,
+    potential_dst_id: EntryId,
+    inner: &mut InnerGraph<N>,
+  ) -> Result<bool, N::Error> {
+    let mut counter = 0;
+    loop {
+      match inner.detect_cycle(src_id, potential_dst_id) {
+        CycleType::Permanent => return Ok(true),
+        CycleType::None => return Ok(false),
+        CycleType::ContainingDirtyNodes(dirty_nodes) => {
+          counter += 1;
+          if counter > 5 {
+            warn!("Couldn't remove cycle containing dirty nodes after {} attempts; nodes in cycle: {:?}", counter, dirty_nodes);
+            return Err(N::Error::cyclic());
+          }
+          inner.invalidate_from_roots(|node| dirty_nodes.contains(node));
+        }
+      }
+    }
+  }
```

Comment:
This could also return its own enum instead of a `Result<bool, _>`, which seems like it would be easier to write and read. I assume that parameterizing the search depth instead of hardcoding it to `5` will be done before merging.

Could there be a very brief description of why we expect invalidating the graph from roots is expected to terminate here? I am not sure I understand when this would or wouldn't happen.

Reply:
@stuhood described some fixed for `detect_cycle()` in https://github.com/pantsbuild/pants/pull/7583#discussion_r276402226, which is relevant to the question.